### DISCUSSION TITLE
How to log unhandled exceptions?

### DISCUSSION THREAD

**Discussion by rubenferreira97** [fastify/fastify#3578] [Q&A] [answered]
I am trying to log unhandled exceptions and reply with a 500 Internal Server Error so I can check and correct bugs later. 
To accomplish this I setup the following Error Handler:
```
fastify.setErrorHandler((error, _, reply) => {
  fastify.log.error(error);
  reply.status(500).send('Oops! 500 Internal Server Error.');
});
```
These are my test endpoints:
```
fastify.get('/', async (_, __) => {
  throw new Error('Test');
});

fastify.get('/secret', async (_, __) => {
  reply.status(401).send('Unauthorized');
});
```

Fastify setErrorHandler catches all errors (even explicit) and not only exceptions. So setErrorHandler is called even when I hit /secret. 
Is there a better approach then checking the type of the error?
Pseudocode:
```
fastify.setErrorHandler((error, _, reply) => {
  if(error instanceof ???) { 
    fastify.log.error(error);
    reply.status(500).send('Oops! 500 Internal Server Error.');
  } else fastify.errorHandler(error, request, response)
});
```

Another problem, I don't know the type (???) or even if Fastify have some specific exception/error for this. I would like a method that handles only exceptions/errors but as far as I know there isn't one.

**Comment by metcoder95** [comment, marked as answer]
The `setErrorHandler` method has the purpose of catching all the errors thrown either intentionally or unintentionally. In auto, fastify will return a `500 Internal Server Error` on every error thrown within the hooks or the main route handler.

If you want to distinguish between errors and exceptions, the best is to use something like [`fastify-sensible`](https://github.com/fastify/fastify-sensible) which allows you to throw meaningful errors in a form of proper HTTP error codes (either client or server ones). For exceptions (those errors thrown by something that went wrong unintentionally), fastify already does proper logging of such errors in a meaningful way, so I recommend you don't do too much in that regard or use the `setErrorHandler` method if you want to decorate them or treat them in a special manner.
